File: .codee/backups/DBZLocaleAI/src/DBZLocaleAI.Core/Runtime/ResourcePlanner.cs
using System;
using System.Collections.Generic;
using System.Diagnostics;

namespace DBZLocaleAI.Core.Runtime
{
    /// <summary>
    /// DBZS Header: Resource Planner
    /// Berechnet optimale Ressourcen-Konfiguration für Modelle
    /// Inspiriert von DBZS Codee Phase 1
    /// Optimiert für Nvidia GTX 1650 (4GB VRAM)
    /// </summary>
    public class ResourcePlanner
    {
        private const long BytesPerMb = 1024 * 1024;
        private const long MinSafetyReserveBytes = 512 * BytesPerMb;
        private const long LowVramSafetyReserveBytes = 768 * BytesPerMb;
        private const long LowVramThresholdBytes = (long)(4.5 * 1024 * BytesPerMb);
        private const int GpuLayerStep = 4;

        // Modell-Größen-Buckets (Schätzungen in Bytes)
        private static readonly Dictionary<string, long> ModelSizeBuckets = new()
        {
            { "70b", 40_000_000_000 },
            { "34b", 20_000_000_000 },
            { "13b", 8_000_000_000 },
            { "8b", 5_500_000_000 },
            { "7b", 5_000_000_000 },
            { "4b", 3_000_000_000 },
            { "3b", 2_500_000_000 },
            { "2b", 1_600_000_000 },
            { "1.5b", 1_600_000_000 },
            { "1b", 1_200_000_000 },
        };

        private const long DefaultModelBytes = 2_000_000_000;

        // KV-Cache-Größe relativ zu F16
        private static readonly Dictionary<string, float> CacheTypeRelativeToF16 = new()
        {
            { "f16", 1.0f },
            { "q8_0", 0.5f },
            { "q4_0", 0.25f },
        };

        // Profil-Kontext-Größen
        private static readonly Dictionary<SlotProfile, int> ProfileContextSizes = new()
        {
            { SlotProfile.FastGpu, 2048 },
            { SlotProfile.QualityCpu, 4096 },
            { SlotProfile.Utility, 2048 },
            { SlotProfile.OrchestratorCpu, 4096 },
        };

        /// <summary>
        /// Berechnet einen Resource-Plan für ein Modell
        /// </summary>
        public ResourcePlan PlanResources(
            string modelId,
            string modelName,
            SlotProfile profile,
            HardwareInfo hardware)
        {
            var warnings = new List<string> { "kv_cache_estimate_approximate" };
            var contextSize = ProfileContextSizes.GetValueOrDefault(profile, 4096);
            var batchSize = profile == SlotProfile.Utility ? 256 : 512;
            var threads = Math.Max(4, hardware.CpuThreads / 2);

            var modelBytes = EstimateModelBytes(modelId, modelName);
            var availableVramBytes = hardware.GpuVramBytes;
            var safetyReserveBytes = CalculateSafetyReserve(availableVramBytes);

            // GPU-Layer Berechnung
            var gpuLayers = 0;
            if (profile == SlotProfile.FastGpu && hardware.GpuVramBytes > 0)
            {
                gpuLayers = ChooseGpuLayers(
                    modelBytes,
                    contextSize,
                    batchSize,
                    hardware.GpuVramBytes,
                    safetyReserveBytes);
            }
            else if (profile == SlotProfile.QualityCpu)
            {
                gpuLayers = 0; // CPU-only
            }

            // Geschätzte VRAM-Nutzung
            var estimatedKvCacheBytes = EstimateKvCacheBytes(contextSize, "q8_0", "q8_0");
            var estimatedComputeBufferBytes = EstimateComputeBufferBytes(batchSize);
            var gpuRatio = gpuLayers > 0 ? Math.Min(gpuLayers / 32.0f, 1.0f) : 0.0f;
            var estimatedTotalVramBytes = gpuLayers > 0
                ? (long)(modelBytes * gpuRatio) + estimatedKvCacheBytes + estimatedComputeBufferBytes
                : 0;

            // Hardware-Mode bestimmen
            var hardwareMode = gpuLayers == 0
                ? "cpu"
                : (gpuLayers >= EstimateMaxGpuLayers(hardware.GpuVramBytes) ? "gpu" : "hybrid");

            // Warnungen hinzufügen
            if (hardware.GpuVramBytes == 0 && profile != SlotProfile.QualityCpu)
            {
                warnings.Add("no_gpu_detected_forced_cpu");
            }

            if (estimatedTotalVramBytes > availableVramBytes - safetyReserveBytes)
            {
                warnings.Add("estimated_vram_exceeds_safety_reserve");
            }

            return new ResourcePlan
            {
                ModelId = modelId,
                SlotId = profile.ToString(),
                ContextSize = contextSize,
                GpuLayers = gpuLayers,
                Threads = threads,
                BatchSize = batchSize,
                EstimatedVramBytes = estimatedTotalVramBytes,
                EstimatedRamBytes = modelBytes + estimatedKvCacheBytes,
                HardwareMode = hardwareMode,
                Warnings = warnings,
            };
        }

        /// <summary>
        /// Schätzt die Modell-Größe basierend auf dem Namen
        /// </summary>
        private long EstimateModelBytes(string modelId, string modelName)
        {
            var searchString = $"{modelId} {modelName}".ToLower();

            foreach (var (token, size) in ModelSizeBuckets)
            {
                if (searchString.Contains(token))
                {
                    return size;
                }
            }

            return DefaultModelBytes;
        }

        /// <summary>
        /// Berechnet die KV-Cache-Größe
        /// </summary>
        private long EstimateKvCacheBytes(int contextSize, string cacheTypeK, string cacheTypeV)
        {
            var baseMb = Math.Max(256, Math.Min(4096, contextSize / 4));
            var relativeK = CacheTypeRelativeToF16.GetValueOrDefault(cacheTypeK, 1.0f);
            var relativeV = CacheTypeRelativeToF16.GetValueOrDefault(cacheTypeV, 1.0f);
            var relative = (relativeK + relativeV) / 2.0f;

            return (long)(baseMb * relative * BytesPerMb);
        }

        /// <summary>
        /// Schätzt die Compute-Buffer-Größe
        /// </summary>
        private long EstimateComputeBufferBytes(int batchSize)
        {
            return Math.Max(256 * BytesPerMb, (long)(batchSize * 1.5 * BytesPerMb));
        }

        /// <summary>
        /// Berechnet die Sicherheits-Reserve
        /// </summary>
        private long CalculateSafetyReserve(long availableVramBytes)
        {
            if (availableVramBytes == 0)
                return MinSafetyReserveBytes;

            var reserve = Math.Max(MinSafetyReserveBytes, (long)(availableVramBytes * 0.15));

            if (availableVramBytes <= LowVramThresholdBytes)
            {
                reserve = Math.Max(reserve, LowVramSafetyReserveBytes);
            }

            return reserve;
        }

        /// <summary>
        /// Wählt die optimale Anzahl von GPU-Layers
        /// </summary>
        private int ChooseGpuLayers(
            long modelBytes,
            int contextSize,
            int batchSize,
            long availableVramBytes,
            long safetyReserveBytes)
        {
            var maxLayers = EstimateMaxGpuLayers(availableVramBytes);

            if (maxLayers <= 0)
                return 0;

            var computeBufferBytes = EstimateComputeBufferBytes(batchSize);
            var kvCacheBytes = EstimateKvCacheBytes(contextSize, "q8_0", "q8_0");

            var layers = maxLayers;
            while (layers > 0)
            {
                var gpuRatio = Math.Min(layers / 32.0f, 1.0f);
                var total = (long)(modelBytes * gpuRatio) + kvCacheBytes + computeBufferBytes;

                if (total <= availableVramBytes - safetyReserveBytes)
                {
                    return layers;
                }

                layers -= GpuLayerStep;
            }

            return 0;
        }

        /// <summary>
        /// Schätzt die maximale Anzahl von GPU-Layers basierend auf VRAM
        /// </summary>
        private int EstimateMaxGpuLayers(long vramBytes)
        {
            if (vramBytes == 0)
                return 0;

            var vramGb = vramBytes / (1024.0 * 1024 * 1024);

            // Heuristik für Nvidia GPUs
            if (vramGb >= 24)
                return 80;
            if (vramGb >= 16)
                return 60;
            if (vramGb >= 12)
                return 40;
            if (vramGb >= 8)
                return 32;
            if (vramGb >= 6)
                return 24;
            if (vramGb >= 4)
                return 16; // GTX 1650 (4GB)
            if (vramGb >= 2)
                return 8;

            return 4;
        }

        /// <summary>
        /// Reduziert den Resource-Plan bei OOM-Fehler
        /// </summary>
        public ResourcePlan ReduceForOom(ResourcePlan previousPlan, int attempt)
        {
            var reductionFactor = Math.Pow(0.5, attempt);
            var newGpuLayers = Math.Max(0, (int)(previousPlan.GpuLayers * reductionFactor));

            if (newGpuLayers == previousPlan.GpuLayers && newGpuLayers > 0)
            {
                newGpuLayers = Math.Max(0, newGpuLayers - GpuLayerStep);
            }

            var warnings = new List<string>(previousPlan.Warnings);
            warnings.Add($"oom_retry_attempt_{attempt}_reduced_gpu_layers_to_{newGpuLayers}");

            if (newGpuLayers == 0)
            {
                warnings.Add("oom_retry_forced_cpu_only");
            }

            var gpuRatio = newGpuLayers > 0 ? Math.Min(newGpuLayers / 32.0f, 1.0f) : 0.0f;
            var estimatedTotalVramBytes = newGpuLayers > 0
                ? (long)(previousPlan.EstimatedVramBytes * gpuRatio)
                : 0;

            return new ResourcePlan
            {
                ModelId = previousPlan.ModelId,
                SlotId = previousPlan.SlotId,
                ContextSize = previousPlan.ContextSize,
                GpuLayers = newGpuLayers,
                Threads = previousPlan.Threads,
                BatchSize = previousPlan.BatchSize,
                EstimatedVramBytes = estimatedTotalVramBytes,
                EstimatedRamBytes = previousPlan.EstimatedRamBytes,
                HardwareMode = newGpuLayers == 0 ? "cpu" : "hybrid",
                Warnings = warnings,
            };
        }
    }
}

File: backend/app/runtime/resource_planner.py
"""Runtime Resource Planner (Phase 1).

Computes a hardware-fit RuntimeResourcePlan for a model/slot instead of the
previous approach of blindly setting `n_gpu_layers=99` or relying on a single
slot-hardcoded override (see docs/RUNTIME_RESOURCE_AUDIT.md). Reuses the
model-size-aware VRAM math already present in
`app.models.profile_service.ProfileService.estimate_resource_usage_for_profile`
instead of reinventing it, and extends its size buckets to cover larger models.
"""
from __future__ import annotations

from app.models.schemas import IndexedModel
from app.runtime.gpu_detect import GpuInfo, _estimate_gpu_layers
from app.runtime.schemas import RuntimeProfileName, RuntimeResourcePlan, RuntimeSlotId

# Extends the name-substring buckets in profile_service.py::estimate_resource_usage_for_profile
# (which only covers 7b/3b/2b/1.5b) with larger tiers, since this planner must also size models
# the profile estimator was never asked to handle.
_SIZE_BUCKETS_BYTES: list[tuple[str, int]] = [
    ("70b", 40_000_000_000),
    ("34b", 20_000_000_000),
    ("13b", 8_000_000_000),
    ("8b", 5_500_000_000),
    ("7b", 5_000_000_000),
    ("4b", 3_000_000_000),
    ("3b", 2_500_000_000),
    ("2b", 1_600_000_000),
    ("1.5b", 1_600_000_000),
]
_DEFAULT_MODEL_BYTES = 2_000_000_000

# KV cache size relative to f16 (llama.cpp's q8_0 cache is ~half f16, q4_0 ~quarter).
_CACHE_TYPE_RELATIVE_TO_F16: dict[str, float] = {"f16": 1.0, "q8_0": 0.5, "q4_0": 0.25}

_BYTES_PER_MB = 1024 * 1024
_MIN_SAFETY_RESERVE_BYTES = 512 * _BYTES_PER_MB
_LOW_VRAM_SAFETY_RESERVE_BYTES = 768 * _BYTES_PER_MB
_LOW_VRAM_THRESHOLD_BYTES = int(4.5 * 1024 * _BYTES_PER_MB)
_GPU_LAYER_STEP = 4

_PROFILE_CONTEXT_SIZE: dict[str, int] = {
    "fast": 2048,
    "balanced": 4096,
    "large_context": 16384,
    "cpu_safe": 4096,
    "hybrid": 4096,
}


def _estimate_model_bytes(model: IndexedModel) -> int:
    if model.size_bytes:
        return model.size_bytes
    haystack = f"{model.id} {model.name}".lower()
    for token, size in _SIZE_BUCKETS_BYTES:
        if token in haystack:
            return size
    return _DEFAULT_MODEL_BYTES


def _estimate_kv_cache_bytes(context_size: int, cache_type_k: str, cache_type_v: str) -> int:
    # Coarse per-token proxy, mirroring profile_service.py's existing
    # `cache_size_mb = max(256, min(4096, context_size // 4))` heuristic — the
    # model index does not carry per-model layer/head-count metadata needed
    # for an exact formula, so callers must treat this as approximate
    # (see the "kv_cache_estimate_approximate" warning added by plan()).
    base_mb = max(256, min(4096, context_size // 4))
    relative = (
        _CACHE_TYPE_RELATIVE_TO_F16.get(cache_type_k, 1.0)
        + _CACHE_TYPE_RELATIVE_TO_F16.get(cache_type_v, 1.0)
    ) / 2.0
    return int(base_mb * relative * _BYTES_PER_MB)


def _estimate_compute_buffer_bytes(batch_size: int) -> int:
    return max(256 * _BYTES_PER_MB, int(batch_size * 1.5 * _BYTES_PER_MB))


def _safety_reserve_bytes(available_vram_bytes: int | None) -> int:
    if available_vram_bytes is None:
        return _MIN_SAFETY_RESERVE_BYTES
    reserve = max(_MIN_SAFETY_RESERVE_BYTES, round(available_vram_bytes * 0.15))
    if available_vram_bytes <= _LOW_VRAM_THRESHOLD_BYTES:
        reserve = max(reserve, _LOW_VRAM_SAFETY_RESERVE_BYTES)
    return reserve


def _fits_in_vram(total_bytes: int, available_vram_bytes: int | None, safety_reserve_bytes: int) -> bool:
    if available_vram_bytes is None:
        return True
    return total_bytes <= available_vram_bytes - safety_reserve_bytes


def _choose_gpu_layers(
    *,
    model_bytes: int,
    context_size: int,
    batch_size: int,
    cache_type_k: str,
    cache_type_v: str,
    gpu: GpuInfo | None,
    available_vram_bytes: int | None,
    safety_reserve_bytes: int,
    max_layers_hint: int | None = None,
) -> int:
    if gpu is None:
        return 0

    ceiling = max_layers_hint if max_layers_hint is not None else _estimate_gpu_layers(gpu.vram_mb)
    if ceiling <= 0:
        return 0

    compute_buffer_bytes = _estimate_compute_buffer_bytes(batch_size)
    kv_cache_bytes = _estimate_kv_cache_bytes(context_size, cache_type_k, cache_type_v)

    layers = ceiling
    while layers > 0:
        gpu_ratio = min(layers / 32.0, 1.0)
        total = int(model_bytes * gpu_ratio) + kv_cache_bytes + compute_buffer_bytes
        if _fits_in_vram(total, available_vram_bytes, safety_reserve_bytes):
            return layers
        layers -= _GPU_LAYER_STEP

    return 0


class RuntimeResourcePlanner:
    def __init__(self) -> None:
        pass

    def plan(
        self,
        model: IndexedModel,
        *,
        slot_id: RuntimeSlotId,
        gpu: GpuInfo | None,
        requested_profile: RuntimeProfileName | None = None,
        custom_overrides: dict[str, object] | None = None,
        mmproj_bytes: int = 0,
    ) -> RuntimeResourcePlan:
        warnings: list[str] = ["kv_cache_estimate_approximate"]
        overrides = custom_overrides or {}
        profile = requested_profile or "balanced"

        context_size = int(overrides.get("context_size", _PROFILE_CONTEXT_SIZE.get(profile, 4096)))
        batch_size = int(overrides.get("batch_size", 512 if profile != "hybrid" else 384))
        micro_batch_size = int(overrides.get("micro_batch_size", min(128, batch_size)))
        parallel = int(overrides.get("parallel", 1))
        threads = int(overrides.get("threads", 4))
        default_cache = "q8_0" if profile == "hybrid" else "f16"
        cache_type_k = str(overrides.get("cache_type_k", default_cache))
        cache_type_v = str(overrides.get("cache_type_v", default_cache))

        available_vram_bytes = (gpu.vram_mb * _BYTES_PER_MB) if gpu else None
        safety_reserve_bytes = _safety_reserve_bytes(available_vram_bytes)
        model_bytes = _estimate_model_bytes(model)
        # Plan 15, Phase 5: a loaded MMProj (vision projector) occupies VRAM
        # alongside the base model's layers. Budget for it up front by shrinking
        # the VRAM available for gpu_layers selection, so a dual-mode vision
        # start on vision_gpu doesn't silently eat into the safety reserve.
        available_vram_bytes_for_layers = (
            max(0, available_vram_bytes - mmproj_bytes) if available_vram_bytes is not None else None
        )

        if profile == "cpu_safe":
            gpu_layers = 0
        elif "gpu_layers" in overrides:
            gpu_layers = int(overrides["gpu_layers"])
        else:
            max_layers_hint = None
            if profile == "fast" and gpu is not None:
                max_layers_hint = _estimate_gpu_layers(gpu.vram_mb)
            elif profile == "hybrid" and gpu is not None:
                # Partial offload: keep headroom for KV + driver on 4GB-class cards.
                max_layers_hint = max(4, _estimate_gpu_layers(gpu.vram_mb) // 2)
            elif profile == "large_context" and gpu is not None:
                # Larger context reserves more room for KV cache, so bias toward
                # fewer GPU layers than the VRAM tier's default recommendation.
                max_layers_hint = max(0, _estimate_gpu_layers(gpu.vram_mb) // 2)

            gpu_layers = _choose_gpu_layers(
                model_bytes=model_bytes,
                context_size=context_size,
                batch_size=batch_size,
                cache_type_k=cache_type_k,
                cache_type_v=cache_type_v,
                gpu=gpu,
                available_vram_bytes=available_vram_bytes_for_layers,
                safety_reserve_bytes=safety_reserve_bytes,
                max_layers_hint=max_layers_hint,
            )
            if profile == "hybrid" and gpu is not None and gpu_layers == 0:
                # Prefer a tiny offload over pure CPU when the user asked for hybrid.
                gpu_layers = min(4, _estimate_gpu_layers(gpu.vram_mb))

        estimated_kv_cache_bytes = _estimate_kv_cache_bytes(context_size, cache_type_k, cache_type_v)
        estimated_compute_buffer_bytes = _estimate_compute_buffer_bytes(batch_size)
        gpu_ratio = min(gpu_layers / 32.0, 1.0) if gpu_layers else 0.0
        estimated_total_vram_bytes = (
            int(model_bytes * gpu_ratio) + estimated_kv_cache_bytes + estimated_compute_buffer_bytes + mmproj_bytes
            if gpu_layers > 0
            else 0
        )

        if gpu_layers == 0:
            hardware_mode = "cpu"
        elif gpu is not None and gpu_layers >= _estimate_gpu_layers(gpu.vram_mb):
            hardware_mode = "gpu"
        else:
            hardware_mode = "hybrid"

        if gpu is None and profile != "cpu_safe" and "gpu_layers" not in overrides:
            warnings.append("no_gpu_detected_forced_cpu")
        if available_vram_bytes is not None and estimated_total_vram_bytes > available_vram_bytes - safety_reserve_bytes:
            warnings.append("estimated_vram_exceeds_safety_reserve")

        return RuntimeResourcePlan(
            model_id=model.id,
            slot_id=slot_id,
            context_size=context_size,
            reserved_output_tokens=int(overrides.get("reserved_output_tokens", round(context_size * 0.22))),
            reserved_tool_tokens=int(overrides.get("reserved_tool_tokens", round(context_size * 0.07))),
            gpu_layers=gpu_layers,
            batch_size=batch_size,
            micro_batch_size=micro_batch_size,
            parallel=parallel,
            threads=threads,
            cache_type_k=cache_type_k,  # type: ignore[arg-type]
            cache_type_v=cache_type_v,  # type: ignore[arg-type]
            estimated_model_bytes=model_bytes,
            estimated_kv_cache_bytes=estimated_kv_cache_bytes,
            estimated_compute_buffer_bytes=estimated_compute_buffer_bytes,
            estimated_total_vram_bytes=estimated_total_vram_bytes,
            available_vram_bytes=available_vram_bytes,
            safety_reserve_bytes=safety_reserve_bytes,
            hardware_mode=hardware_mode,
            warnings=warnings,
            mmproj_bytes=mmproj_bytes,
        )

    def reduce_for_oom(self, previous_plan: RuntimeResourcePlan, attempt: int) -> RuntimeResourcePlan:
        """Return a copy of previous_plan with fewer GPU layers, for OOM retry.

        `attempt` is the retry count so far (1, 2, 3, ...); the reduction gets
        more aggressive each attempt. Never silently drops to gpu_layers=0
        without recording why in `warnings`.
        """
        reduction_factor = 0.5 ** attempt
        new_gpu_layers = max(0, int(previous_plan.gpu_layers * reduction_factor))
        if new_gpu_layers == previous_plan.gpu_layers and new_gpu_layers > 0:
            new_gpu_layers = max(0, new_gpu_layers - _GPU_LAYER_STEP)

        warnings = list(previous_plan.warnings)
        warnings.append(f"oom_retry_attempt_{attempt}_reduced_gpu_layers_to_{new_gpu_layers}")
        if new_gpu_layers == 0:
            warnings.append("oom_retry_forced_cpu_only")

        gpu_ratio = min(new_gpu_layers / 32.0, 1.0) if new_gpu_layers else 0.0
        # Carry the original mmproj_bytes forward: the projector is still loaded
        # even when we reduce gpu_layers, so it still consumes VRAM.  Including
        # it keeps estimated_total_vram_bytes honest across retries.
        mmproj = previous_plan.mmproj_bytes
        estimated_total_vram_bytes = (
            int(previous_plan.estimated_model_bytes * gpu_ratio)
            + previous_plan.estimated_kv_cache_bytes
            + previous_plan.estimated_compute_buffer_bytes
            + mmproj
            if new_gpu_layers > 0
            else 0
        )

        return previous_plan.model_copy(
            update={
                "gpu_layers": new_gpu_layers,
                "estimated_total_vram_bytes": estimated_total_vram_bytes,
                "hardware_mode": "cpu" if new_gpu_layers == 0 else "hybrid",
                "warnings": warnings,
                # mmproj_bytes unchanged — projector residency is independent of
                # gpu_layers: llama-server keeps it mapped in VRAM regardless.
            }
        )
